src/mscorlib/src/System/String.cs

```diff
@@ -2241,15 +2218,42 @@ public static int Compare(String strA, String strB, StringComparison comparisonT
         //
         [Pure]
         [System.Security.SecuritySafeCritical]  // auto-generated
-        public static int CompareOrdinal(String strA, int indexA, String strB, int indexB, int length) {
-           if (strA == null || strB == null) {
-                if ((Object)strA==(Object)strB) { //they're both null;
+        public static int CompareOrdinal(String strA, int indexA, String strB, int indexB, int length)
+        {
+            if (strA == null || strB == null)
+            {
+                if (object.ReferenceEquals(strA, strB))
+                {
+                    // They're both null
                     return 0;
                 }
 
-                return (strA==null)? -1 : 1; //-1 if A is null, 1 if B is null.
+                return strA == null ? -1 : 1;
             }
 
+            // We can return early if length == 0 or (object)strA == strB && indexA == indexB.
+            // However, since we didn't previously apply this optimization, we have to make sure
+            // first that an exception wouldn't have been thrown had we continued. Otherwise,
+            // this would be a breaking change.
+
+            // We're allowed to return early for nulls above, e.g. CompareOrdinal(null, -1, null, -1, -1)
+            // will not throw despite the invalid parameters, since this was the original behavior of the method.
+
+            bool exceptionImminent = length < 0 || (uint)indexA > (uint)strA.Length || (uint)indexB > (uint)strB.Length;
```

Comment:
How about moving this code to within the other if statement below (switch order of if statements). That way we only pay for it if a short circuit is in question.